test/e2e/puppeteer.js

```diff
@@ -1,195 +1,257 @@
-console.null = () => {};
 
+let browser;
 
 /* Launch server */
 
-const server = http.createServer( ( req, resp ) => handler( req, resp ) );
-server.listen( port, async () => await pup );
-server.on( 'SIGINT', () => process.exit( 1 ) );
+const server = http.createServer( handler );
+server.listen( port, main );
+server.on( 'SIGINT', () => {
+
+	server.close();
+	if ( browser !== undefined ) browser.close();
+	process.exit( 1 );
+
+} );
+
+async function downloadLatestChromium() {
+
+	const browserFetcher = puppeteer.createBrowserFetcher();
+
+	const lastRevisionURL = LAST_REVISION_URLS[ browserFetcher.platform() ];
+	const revision = await ( await fetch( lastRevisionURL ) ).text();
+
+	const revisionInfo = browserFetcher.revisionInfo( revision );
+	if ( revisionInfo.local === true ) {
+
+		console.log( 'Latest Chromium has been already downloaded.' );
+		return revisionInfo;
+
+	} else {
+
+		console.log( 'Downloading latest Chromium...' );
+		const revisionInfo = await browserFetcher.download( revision );
+		console.log( 'Downloaded.' );
+		return revisionInfo;
+
+	}
+
+}
 
+async function main() {
 
-/* Launch browser */
+	/* Find files */
+
+	const isMakeScreenshot = process.argv[ 2 ] === '--make';
+
+	const exactList = process.argv.slice( isMakeScreenshot ? 3 : 2 )
+		.map( f => f.replace( '.html', '' ) )
+		.map( regexify );
+
+	const isExactList = exactList.length !== 0;
+
+	const files = ( await fs.readdir( './examples' ) )
+		.filter( s => s.slice( - 5 ) === '.html' && s !== 'index.html' )
+		.map( s => s.slice( 0, s.length - 5 ) )
+		.filter( f => isExactList ? exactList.some( r => r.test( f ) ) : ! exceptionList.some( r => r.test( f ) ) );
+
+	if ( isExactList ) {
+
+		for ( let regexp of exactList ) {
+
+			if ( ! files.some( f => regexp.test( f ) ) ) {
+
+				console.log( `Warning! Unrecognised example name: ${ unregexify( regexp ) }` );
+
+			}
+
+		}
+
+	}
+
+	/* Download browser */
 
-const pup = puppeteer.launch( {
-	headless: ! process.env.VISIBLE,
-	args: [
-		'--use-gl=swiftshader',
-		'--no-sandbox',
-		'--enable-surface-synchronization'
-	]
-} ).then( async browser => {
+	const { executablePath } = await downloadLatestChromium();
 
+	const flags = ( process.platform === 'darwin' || process.platform === 'win32' ) ? [ '--use-angle', '--enable-unsafe-webgpu' ] : [];
+
+	/* Launch browser */
+
+	browser = await puppeteer.launch( {
+		executablePath,
+		headless: ! process.env.VISIBLE,
+		args: flags
+	} );
 
 	/* Prepare page */
 
 	const page = ( await browser.pages() )[ 0 ];
 	await page.setViewport( { width: width * viewScale, height: height * viewScale } );
 
-	const cleanPage = fs.readFileSync( 'test/e2e/clean-page.js', 'utf8' );
-	const injection = fs.readFileSync( 'test/e2e/deterministic-injection.js', 'utf8' );
-	await page.evaluateOnNewDocument( injection );
-
-	const threeJsBuild = fs.readFileSync( 'build/three.module.js', 'utf8' )
-		.replace( /Math\.random\(\) \* 0xffffffff/g, 'Math._random() * 0xffffffff' );
```

Comment:
Probably I just needed to tag this lines. But I hope it become more clear why it is useful.

Reply:
I still don't see a point in changing **deterministic UUIDs** with injected Math.random to **non-deterministic** with native random Math._random.

Reply:
I am used `.replace( /Math\.random\(\) \* 0xffffffff/g, 'crypto.getRandomValues(new Uint32Array(1))[0]' );` before, but it was changed to non-determenistic one.

Fun fact that it’s working too, because this is only for UUID generation, that should be just different all the time, and that’s all.

If you found solution with `crypto` more clear you can use it or leave comment there.

Reply:
`crypto.getRandomValues(new Uint32Array(1))[0]` is also non-deterministic.

I don't see why we should use non-determinstic random.

Reply:
No, I mean that the problem is in every call of the Math.random? This regex only "fixes" it in UUIDs but not in other places.

Reply:
Seems you right. It can be `.replace( /Math\.random\(\)/g, 'someNewDeterministicRandom()' );`.
But not `.replace( /Math\.random\(\)/g, 'Math._random()' );`.

Reply:
I don't see a way to fix this problem without creating separate independent generators for each kind of task (UUIDs generation, color generation, etc)...

Reply:
Such changes in Math.random calls order happen quite rarely.
I think we can just remake screenshots in such rare cases.

Reply:
Don’t see a reason to remove this two lines of code. You just recreating lot of screenshots without good reason now. Also in future any code that creating additional texture in example/core will change **example itself**.

The fact that UUID using the same `math.random` function is just a coincidence and was leaved like that because nobody can suggest something more performant or new, UUID can be based on a time. All other `math.random` in core not used so frequently as in UUID.

Reply:
OK, I see... OK, will return that.